File: go/test/graph_ql_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/fussy-api-documentation-sdk/go"
	"github.com/voxgig-sdk/fussy-api-documentation-sdk/go/core"

	vs "github.com/voxgig-sdk/fussy-api-documentation-sdk/go/utility/struct"
)

func TestGraphQlEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.GraphQl(nil)
		if ent == nil {
			t.Fatal("expected non-nil GraphQlEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"graph_ql": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.GraphQl(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.MakeConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.GraphQl(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := graph_qlBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "graph_ql." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set FUSSY_API_DOCUMENTATION_TEST_GRAPH_QL_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		graphQlRef01Ent := client.GraphQl(nil)
		graphQlRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "graph_ql"}, setup.data), "graph_ql_ref01"))

		graphQlRef01DataResult, err := graphQlRef01Ent.Create(graphQlRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		graphQlRef01Data = core.ToMapAny(entityData(graphQlRef01DataResult))
		if graphQlRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}

		// LIST
		graphQlRef01Match := map[string]any{}

		graphQlRef01ListResult, err := graphQlRef01Ent.List(graphQlRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, graphQlRef01ListOk := graphQlRef01ListResult.([]any)
		if !graphQlRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", graphQlRef01ListResult)
		}

	})
}

func graph_qlBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "graph_ql", "GraphQlTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read graph_ql test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse graph_ql test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"graph_ql01", "graph_ql02", "graph_ql03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("FUSSY_API_DOCUMENTATION_TEST_GRAPH_QL_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"FUSSY_API_DOCUMENTATION_TEST_GRAPH_QL_ENTID": idmap,
		"FUSSY_API_DOCUMENTATION_TEST_LIVE":      "FALSE",
		"FUSSY_API_DOCUMENTATION_TEST_EXPLAIN":   "FALSE",
		"FUSSY_API_DOCUMENTATION_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["FUSSY_API_DOCUMENTATION_TEST_GRAPH_QL_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["FUSSY_API_DOCUMENTATION_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["FUSSY_API_DOCUMENTATION_APIKEY"],
			},
			extra,
		})
		client = sdk.NewFussyApiDocumentationSDK(core.ToMapAny(mergedOpts))
	}

	live := env["FUSSY_API_DOCUMENTATION_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["FUSSY_API_DOCUMENTATION_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: rb/test/graph_ql_entity_test.rb
# GraphQl entity test

require "minitest/autorun"
require "json"
require_relative "../FussyApiDocumentation_sdk"
require_relative "runner"

class GraphQlEntityTest < Minitest::Test
  def test_create_instance
    testsdk = FussyApiDocumentationSDK.test(nil, nil)
    ent = testsdk.GraphQl(nil)
    assert !ent.nil?
  end

  # Feature #4: the entity stream(action, ...) method runs the op pipeline and
  # returns an Enumerator over result items. With the streaming feature active
  # it yields the feature's incremental output; otherwise it falls back to the
  # materialised list so stream always yields.
  def test_stream
    seed = {
      "entity" => {
        "graph_ql" => {
          "s1" => { "id" => "s1" },
          "s2" => { "id" => "s2" },
          "s3" => { "id" => "s3" },
        },
      },
    }

    # Fallback: streaming inactive -> yields the materialised list items.
    base = FussyApiDocumentationSDK.test(seed, nil)
    seen = base.GraphQl(nil).stream("list", nil, nil).to_a
    assert_equal 3, seen.length

    # Inbound: streaming active -> yields each item from the feature.
    cfg = FussyApiDocumentationConfig.make_config
    if cfg["feature"].is_a?(Hash) && cfg["feature"].key?("streaming")
      sdk = FussyApiDocumentationSDK.test(seed, { "feature" => { "streaming" => { "active" => true } } })
      got = []
      sdk.GraphQl(nil).stream("list", nil, nil).each do |item|
        if item.is_a?(Array)
          got.concat(item)
        else
          got << item
        end
      end
      assert_equal 3, got.length
    end
  end

  def test_basic_flow
    setup = graph_ql_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create", "list"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "graph_ql." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set FUSSY_API_DOCUMENTATION_TEST_GRAPH_QL_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    graph_ql_ref01_ent = client.GraphQl(nil)
    graph_ql_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.graph_ql"), "graph_ql_ref01"))

    graph_ql_ref01_data_result = graph_ql_ref01_ent.create(graph_ql_ref01_data, nil)
    graph_ql_ref01_data = Helpers.to_map(graph_ql_ref01_data_result.respond_to?(:data_get) ? graph_ql_ref01_data_result.data_get : graph_ql_ref01_data_result)
    assert !graph_ql_ref01_data.nil?

    # LIST
    graph_ql_ref01_match = {}

    graph_ql_ref01_list_result = graph_ql_ref01_ent.list(graph_ql_ref01_match, nil)
    assert graph_ql_ref01_list_result.is_a?(Array)

  end
end

def graph_ql_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "graph_ql", "GraphQlTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = FussyApiDocumentationSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["graph_ql01", "graph_ql02", "graph_ql03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["FUSSY_API_DOCUMENTATION_TEST_GRAPH_QL_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "FUSSY_API_DOCUMENTATION_TEST_GRAPH_QL_ENTID" => idmap,
    "FUSSY_API_DOCUMENTATION_TEST_LIVE" => "FALSE",
    "FUSSY_API_DOCUMENTATION_TEST_EXPLAIN" => "FALSE",
    "FUSSY_API_DOCUMENTATION_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["FUSSY_API_DOCUMENTATION_TEST_GRAPH_QL_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["FUSSY_API_DOCUMENTATION_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["FUSSY_API_DOCUMENTATION_APIKEY"],
      },
      extra || {},
    ])
    client = FussyApiDocumentationSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["FUSSY_API_DOCUMENTATION_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["FUSSY_API_DOCUMENTATION_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
